### PR TITLE
bash,texinfo: turn interactive versions into passthrus

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change

This patchset is a part of a bigger patchset (that needs a bit more work) that adds `getInteractive` to the lib (or maybe not, that is a subject to change) and then uses that to build user-facing environments. I.e. you, as a user, won't need to ask yourself "Do I need specify Y or YInteractive here?" after applying that patchset, you just say "Y" and it will then select the appropriate one for every instance. E.g. `bash` in FHS environments and NixOS shells will be automatically referenced as `bash.interactive`.

###### Notes

- b51ef9ece59be09c466d84b5387e5534d1458d3a is related, but unrelated (it's actually a piece of `doCheck` patchset), bundled here for simplicity.

- I do `let self =` thing here instead of pushing `self` attr from `all-packages.nix` because both `bash` and `texinfo` are used in `stdenv` stages. Doing this the other way like, e.g. python does, needs some really ugly changes to stdenv.

###### Things done

- [X] Writing from a system with that patchset applied. It worked before rebasing onto master.

### CODE REVIEW

File: nixos/modules/config/system-path.nix [hunk 1]
```diff
@@ -11,7 +11,7 @@ let
     [ config.nix.package
       pkgs.acl
       pkgs.attr
-      pkgs.bashInteractive # bash with ncurses support
+      pkgs.bash.interactive # bash with ncurses support
```
Comment (volth) [inline nixos/modules/config/system-path.nix]: as the old names (```bashInteractive```) are going to be broken, maybe it is time to invent a more precise name which would not need to be explained?

```bash.withNcurses``` ?

**Comment by oxij** [comment]
> as the old names (```bashInteractive```) are going to be broken, maybe it is time to invent a more precise name which would not need to be explained? > ```bash.withNcurses``` ?

The problem with `bash.withNcurses` is that it's too specific. As I said in the OP, I want FHS environments and such to select `.interactive` when it's available automatically.

I.e.

- you put `Y` into nativeBuildInputs -> you get `Y.bin/Y.out` (`getBin`)
- you put `Y` into systemPackages    -> you get `Y.interactive/Y.out` (`getInteractive`)
- you can get others by asking for them specifically, e.g. with `Y.out` in `systemPackages`

Using some variant of the `self` trick to provide packages with useful overrides as passthrus may be a good idea, though. If we want users to use attr names, then grouping package variants under the main package makes sense to me. Like `emacs.nox`, `git.full`, etc.

I would make a cleaner mechanism for it, though. Like we can make `callPackage` to supply `self` and `.override` override it, or something. Then most of this hackery wouldn't be needed.

ATM, however, this `let self` hackery is needed or else `texinfo` and `bash` will be rebuilt multiple times in for the `stdenv`, these packages are kinda special.

**Comment by matthewbauer** [comment]
This makes things way too complicated in my opinion. Putting overrides as passthru attrs just hides the problem, not fixes it. I want us to get to a point where we can do:

```
bashInteractive = bash;
texinfoInteractive = texinfo;
```

The bash/bashInteractive & texinfo/texinfoInteractive are just needed for bootstrapping. Besides that there is no reason we can't just make the default package the fully featured thing that everyone wants. Doing this without breaking stdenv is tricky but I think it can be done with overrides. You would just need to override each bash/texinfo instance listed in darwin/linux stdenv with ```.override { interactive = false; }```. If everything else is done correctly, I believe this would work. Need some time to actually implement it though.

**Comment by oxij** [comment]
@matthewbauer

> ```
> bashInteractive = bash;
> texinfoInteractive = texinfo;
> ```

I see, I agree that that would be better than this.

But what are your general thoughts on the ideas expressed in my comment above?

**Comment by oxij** [comment]
New ideas I will implement and test later, this is closed in favor of eventually changing default `texinfo` and `bash` to interactive versions.

**Pull request closed without merging**